### PR TITLE
Temporal reduction

### PR DESCRIPTION
**Pull request by burlen**
adds teca_temporal_reduction which can reduce a time axis to monthly or daily resolution using average, min, or max operators.

### CODE REVIEW

File: alg/teca_temporal_reduction.py [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+            # time point's to iterate between
+            self.t0 = time_point(t[0], units, calendar)
+            self.t1 = time_point(t[-1], units, calendar)
+
+            # current time state
+            self.year = self.t0.year
+            self.month = self.t0.month
+
+        def last_day_of_month(self):
```
Comment (taobrienlbl) [inline alg/teca_temporal_reduction.py:73]: can we use the `calcalcs` `ccs_dpm()` routine in here instead?  Examining `calcalcs.cxx` shows that the logic for computing this is substantially more complicated than just considering leap years.  For example, the 360_day calendar should have 30 days in all months.

File: alg/teca_temporal_reduction.py [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+            self.day = self.t0.day
+
+        def last_day_of_month(self):
+            """
+            get the number of days in the month, with logic for
+            leap years
+            """
+
+            # first entry is -1 because month numbers start at 1 not zero
+            days = [-1, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
```
Comment (taobrienlbl) [inline alg/teca_temporal_reduction.py:170]: same comment here as above: can we use `ccs_dpm()` inside this function instead?

File: python/teca_py_core.i [hunk 3]
```diff
@@ -683,3 +684,38 @@ typedef std::pair<std::shared_ptr<teca_algorithm>, unsigned int> teca_algorithm_
 %ignore teca_index_reduce::operator=;
 %include "teca_index_reduce_fwd.h"
 %include "teca_index_reduce.h"
+
```
Comment (taobrienlbl) [inline python/teca_py_core.i:687]: Following the above comments, I'd suggest also wrapping `calcalcs::ccs_dpm()`

**Comment by burlen** [comment]
@taobrienlbl You are correct about the flaws in my approach, I hadn't fully thought it through. I didn't know about the calcalcs function you pointed out, and as I read it I see that there are a number of other useful functions in there. I agree that using the calcalcs implementations is the way to go. I'll make that change.

File: apps/teca_temporal_reduction.in [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+parser.add_argument('--n_threads', type=int, default=2,
+                    help='Number of threads to use when stremaing the '
+                         'reduction (2)')
+
+parser.add_argument('--verbose', type=int, default=1,
+                    help='enable verbose mode. (1)')
+
+
+args = parser.parse_args()
```
Comment (taobrienlbl) [inline apps/teca_temporal_reduction.in:56]: if the parser fails, all mpi ranks print the failure/help message.  Maybe this needs to be wrapped like:
```python
try:
    args = parser.parse_args()
except:
    if rank == 0:
        raise
```

**Comment by taobrienlbl** [comment]
@burlen -- cool, glad you agree that sounds like an OK way to go.

Btw, I just tested this on NERSC -- it's very fast!  I'm really glad to have this as a new tool!

**Comment by taobrienlbl** [comment]
@burlen -- just noticed that `byte` type netCDF data doesn't appear to be reduced properly with the averaging operator.  My initial guess is that the `+` operation is taking place on byte-type data.  `int` and `byte` types may need to be coerced to `float` for the averaging reduction.

**Comment by burlen** [comment]
> just noticed that `byte` type netCDF data doesn't appear to be reduced properly with the averaging operator.

Oh. Yeah. That makes sense because Byte would have a range of -128 to 127 (or 0 255) so overflow would almost certainly occur. I will fix that...

**Comment by burlen** [comment]
@taobrienlbl I guess for all integer types we'd want to use a floating point type for averaging. Perhaps float for int8 - int32 and double for int64? What do you think?

**Comment by taobrienlbl** [comment]
>   @taobrienlbl I guess for all integer types we'd want to use a floating point type for averaging. Perhaps float for int8 - int32 and double for int64? What do you think?

Yes, I think so too.  I can't think of a use-case where integer division for an average would make sense.

**Comment by burlen** [comment]
@taobrienlbl Would a summation operator for `ar_binary_tag` make more sense than the average operator? Also I'm thinking now the option to specify a different operator for each array would be useful.

**Comment by taobrienlbl** [comment]
that's an interesting thought.  I think that for my immediate uses, the average would be the one I would lean toward.  But I could see how a summation could be useful.  If it's easier to implement than variable coercion, I would say go for it.  Perhaps the summation count could be put in the variable or global metadata?

And I think I would lean toward enforcing one reduction operation per run. Otherwise, I'm concerned the reduction app would become overly complicated. For example, say I want to calculate the monthly average of 6-hourly PSL and I also want to calculate the monthly minimum: if we contain them in the same file, then we would also have to have to map input variables to output+operation names.

Then again, maybe we do want to rename variables to reflect the reduction operation and timescale ('PSL' -> 'PSL_monthly_avg') and also add metadata to the variable to indicate the details of the reduction.  This would then lend itself to an API with options like `--average_arrays PSL, TMQ` and `--max_arrays PSL, TMQ, U850`, which isn't too complicated.

Anyway, maybe we can consider these additions to be new features to add in another PR?

**Comment by burlen** [comment]
> for my immediate uses, the average would be the one I would lean toward.

> maybe we can consider these additions to be new features to add in another PR?

Sounds good

File: alg/teca_temporal_reduction.py (hunk 1 above)
Comment (burlen) [reply alg/teca_temporal_reduction.py:68]: I exposed this functionality in a new function `calcalcs::days_in_month`

**Comment by burlen** [comment]
@taobrienlbl I made the suggested changes, all set for another look

**Pull request merged**